godot-core/src/export.rs

```diff
@@ -0,0 +1,139 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+use crate::{builtin::GodotString, engine::global::PropertyHint};
+
+use godot_ffi as sys;
+
+/// Trait implemented for types that can be used as `#[export]` fields. This creates a copy of the
+/// value, for some type-specific definition of "copy". For example, `Array` and `Gd` are returned
+/// via `Share::share()` instead of copying the actual data.
+pub trait Export: ExportDefaultInfo {
+    /// Creates a copy to be returned from a getter.
+    fn export(&self) -> Self;
+}
+
+/// Trait implemented for types that can be used in an `#[export]` field without specifying a custom `hint`
+/// and `hint_string`.
+pub trait ExportDefaultInfo {
+    /// The property hint to use when exporting properties if a custom one isn't specified.
+    fn default_property_hint() -> PropertyHint {
+        PropertyHint::PROPERTY_HINT_NONE
+    }
+
+    /// The property hint string to use when exporting properties if a custom one isn't specified.
+    fn default_property_hint_string() -> GodotString {
+        GodotString::new()
+    }
+}
+
+impl<T: ExportDefaultInfo> ExportDefaultInfo for Option<T> {
+    fn default_property_hint() -> PropertyHint {
+        T::default_property_hint()
+    }
+
+    fn default_property_hint_string() -> GodotString {
+        T::default_property_hint_string()
+    }
+}
+
+impl<T: Export> Export for Option<T> {
+    fn export(&self) -> Self {
+        self.as_ref().map(|val| val.export())
+    }
+}
+
+// Implement ExportDefaultInfo for some various common types so people can use them without specifying a
+// hint_type and hint_string.
+impl ExportDefaultInfo for () {}
+impl ExportDefaultInfo for String {}
+impl ExportDefaultInfo for i128 {}
+impl ExportDefaultInfo for isize {}
+impl ExportDefaultInfo for u64 {}
+impl ExportDefaultInfo for u128 {}
+impl ExportDefaultInfo for usize {}
```

Comment:
All these types should not be exportable in my opinion.
* `i128`, `u128`, `u64` cannot be represented in Godot, so it's just unnecessary potential for errors.
* `isize` and `usize` have a platform-dependent size, so it makes shipping robust applications harder. Godot also has no equivalent for them.
* `()` can only ever have one value (aka zero information), so what's the idea here?
* `String` is too expensive because it needs copying + reallocation on every access (see [#gdnative/1020](https://github.com/godot-rust/gdnative/pull/1020)).